src/middlewared/middlewared/plugins/cluster_linux/ctdb_private.py

```diff
@@ -0,0 +1,104 @@
+from middlewared.schema import List, Dict, IPAddr, Int
+from middlewared.service import accepts, private, job, CRUDService, ValidationErrors
+import middlewared.sqlalchemy as sa
+from .utils import CTDBConfig, JOB_LOCK
+
+
+PRIVATE_IP_FILE = CTDBConfig.PRIVATE_IP_FILE.value
+
+
+class CtdbPrivateIpModel(sa.Model):
+    __tablename__ = 'ctdb_private_ips'
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+    ip = sa.Column(sa.String(45), nullable=True)
+
+
+class CtdbPrivateIpService(CRUDService):
+
+    class Config:
+        namespace = datastore = 'ctdb.private.ips'
+
+    @private
+    def common_validation(self, data, schema_name, verrors):
+
+        # normalize the data and remove netmask info (if its given to us)
+        # from `ip` keys since it's not used
+        for i in data['private_ips']:
+            i['ip'] = i['ip'].split('/')[0]
+
+        # get the current ips in the cluster
+        cur_data = {'private_ips': [], 'public_ips': []}
+        cur_data['private_ips'] = self.middleware.call_sync('ctdb.private.ips.query')
+        cur_data['public_ips'] = self.middleware.call_sync('ctdb.public.ips.query')
+
+        # need to make sure the new private ip(s) don't already exist in the cluster
+        if m := list(set(i['ip'] for i in data['private_ips']).intersection(set(i['ip'] for i in cur_data['private_ips']))):
+            verrors.add(
+                f'{schema_name}.ip',
+                f'Private IP address(es): {m} already in the cluster.',
+            )
+
+        # need to make sure the new private ip(s) don't already exist in the cluster as public ip(s)
+        if m := list(set(i['ip'] for i in data['private_ips']).intersection(set(i['ip'] for i in cur_data['public_ips']))):
+            verrors.add(
+                f'{schema_name}.ip',
+                f'Private IP address(es): {m} have already been added to the cluster as public IP(s).',
+            )
+
+        verrors.check()
+
+        return data
+
+    @private
+    def write_private_ips_to_ctdb(self):
+
+        data = self.middleware.call_sync('ctdb.private.ips.query')
+
+        with open(PRIVATE_IP_FILE, 'w') as f:
+            for i in map(lambda i: i['ip'], data):
+                f.write(f'{i}\n')
+
+        return data
+
+    @accepts(Dict(
+        'node_create',
+        List('private_ips', unique=True, empty=False, items=[Dict('private_ip', IPAddr('ip'))]),
+    ))
+    @job(lock=JOB_LOCK)
+    def do_create(self, job, data):
```

Comment:
I am still confused by the CRUD Service.

Usually a CRUD Service is when we can create one entity, update it, delete it.

But ion here are are performing a batch create, which seems to defeat the CRUD paradigm since I cant operate over a single entity.